### PR TITLE
8231286: HTML font size too large with high-DPI scaling and W3C_UNIT_LENGTHS

### PR DESCRIPTION
**Pull request by mperktold**
@prsadhuk asked me to take over his pull request #1628, since I filed the bug an suggested this fix.

I thought the current behavior would be buggy, but actually the units are quite precise. I checked the size of a text in font-size of 1 in, and it really was approximately 1 inch. The problem is just that browsers behave differently.

Swing follows the CSS 2.1 spec: https://www.w3.org/TR/CSS21/syndata.html#length-units.
But in CSS 2.2, length units where redefined: https://www.w3.org/TR/CSS22/syndata.html#length-units.
Now `px` is also an absolute unit, and there are constant conversion factors between all absolute units.

The CSS 2.2 spec includes the following two notes regarding this change:

> Note that if the anchor unit is the pixel unit, the physical units might not match their physical measurements. Alternatively if the anchor unit is a physical unit, the pixel unit might not map to a whole number of device pixels.

> Note that this definition of the pixel unit and the physical units differs from previous versions of CSS. In particular, in previous versions of CSS the pixel unit and the physical units were not related by a fixed ratio: the physical units were always tied to their physical measurements while the pixel unit would vary to most closely match the reference pixel. (This change was made because too much existing content relies on the assumption of 96dpi, and breaking that assumption breaks the content.)

So the spec changed the behavior to better support high-DPI devices with existing content, and that is exactly my intention with this PR as well.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8231286](https://bugs.openjdk.java.net/browse/JDK-8231286)

### Issue
 * [JDK-8231286](https://bugs.openjdk.java.net/browse/JDK-8231286): HTML font size too large with high-DPI scaling and W3C_LENGTH_UNITS ⚠️ Title mismatch between PR and JBS.

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2223/head:pull/2223`
`$ git checkout pull/2223`

### CODE REVIEW

File: src/java.desktop/share/classes/javax/swing/JEditorPane.java [hunk 1]
```diff
@@ -1564,7 +1564,7 @@ private void writeObject(ObjectOutputStream s) throws IOException {
 
     /**
      * Key for a client property used to indicate whether
-     * <a href="http://www.w3.org/TR/CSS21/syndata.html#length-units">
+     * <a href="https://www.w3.org/TR/CSS22/syndata.html#length-units">
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/JEditorPane.java:1567]: Does this change require a CSR?

File: test/jdk/javax/swing/text/html/HtmlFontSizeTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+import javax.swing.JFrame;
+import javax.swing.JEditorPane;
+import javax.swing.SwingUtilities;
+import javax.swing.text.Document;
+import javax.swing.text.html.HTMLEditorKit;
+import javax.swing.text.html.StyleSheet;
+
+public class HtmlFontSizeTest {
+    static Rectangle w3cFrameSize, stdFrameSize;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HtmlFontSizeTest.java:42]: These fields are accessed from two different threads. Shall they be declared as `volatile` then?
Declaring each field on its own line would follow the Java Code Conventions.

File: test/jdk/javax/swing/text/html/HtmlFontSizeTest.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    public static void main(String[] args) throws Exception {
+        SwingUtilities.invokeAndWait(() -> {
+            w3cFrameSize = test(true);
+            stdFrameSize = test(false);
+        });
+        System.out.println("frame height with W3C:" + w3cFrameSize);
+        System.out.println("frame height without W3C:" + stdFrameSize);
+
+        float ratio = (float)w3cFrameSize.width/(float)stdFrameSize.width;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HtmlFontSizeTest.java:92]: Spaces around / operator.
Java Code Conventions recommend “casts should be followed by a blank,” however, it's not always followed.
```suggestion
        float ratio = (float)w3cFrameSize.width / (float)stdFrameSize.width;
```

File: test/jdk/javax/swing/text/html/HtmlFontSizeTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        System.out.println("w3cFrameSize.width/stdFrameSize.width " + ratio);
+
+        String str = String.format("%.1g%n",ratio);
+        if (str.compareTo(String.format("%.1g%n",1.3f)) != 0) {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HtmlFontSizeTest.java:96]: “A comma should appear after commas in argument lists.”
```suggestion
        String str = String.format("%.1g%n", ratio);
        if (str.compareTo(String.format("%.1g%n", 1.3f)) != 0) {
```

File: test/jdk/javax/swing/text/html/HtmlFontSizeTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HtmlFontSizeTest.java:2]:
```suggestion
 * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```

**Comment by aivanov-jdk** [comment]
The comment in the bug description says, the subject contains `W3C_UNIT_LENGTHS` instead of `W3C_LENGTH_UNITS`. Shall we edit the bug subject so that it contains the correct constant name?

**Comment by aivanov-jdk** [comment]
> As of now, I did not check the change in the spec, just noted that the quoted text in the PR is the same for both links: CSS 2.1 and CSS 2.2.

The text remains the same but the external link changes from CSS 2.1 to 2.2.
It probably does not change the spec; just updates the link to the more recent spec which Swing will follow after the fix.

**Comment by mrserb** [comment]
> The text remains the same but the external link changes from CSS 2.1 to 2.2.
> It probably does not change the spec; just updates the link to the more recent spec which Swing will follow after the fix.

This is what I mean, the quoted text of the CSS spec is the same for both versions.

**Comment by mperktold** [comment]
> This is what I mean, the quoted text of the CSS spec is the same for both versions.

Oh right I misread that, I really thought it changed, but it didn't.
In that case I can also leave the link as it was, should I revert that change?

**Comment by aivanov-jdk** [comment]
> > This is what I mean, the quoted text of the CSS spec is the same for both versions.
> 
> Oh right I misread that, I really thought it changed, but it didn't.
> In that case I can also leave the link as it was, should I revert that change?

I believe we're talking about javadoc spec: its text has not changed.
CSS spec is updated. Since you follow the rules of CSS 2.2, it's better to link to CSS 2.2.

So, the changes are good as they are. Do I get right, @mrserb?

**Comment by mperktold** [comment]
> 
> 
> ⚠️ @mperktold a branch with the same name as the source branch for this pull request (`master`) is present in the [target repository](https://github.com/openjdk/jdk). If you eventually integrate this pull request then the branch `master` in your [personal fork](https://github.com/mperktold/jdk) will diverge once you sync your personal fork with the upstream repository.
> 
> To avoid this situation, create a new branch for your changes and reset the master branch. You can do this by running the following commands in a local repository for your personal fork. [...]

I am not sure whether I should still do this or not, since there is already some feedback, and I don't know what happens to that.
In particular, I didn't understand if this is just a suggestion or actually required.

@aivanov-jdk if it's not required, do you have any preferences?

**Comment by aivanov-jdk** [comment]
> > ⚠️ @mperktold a branch with the same name as the source branch for this pull request (`master`) is present in the [target repository](https://github.com/openjdk/jdk). If you eventually integrate this pull request then the branch `master` in your [personal fork](https://github.com/mperktold/jdk) will diverge once you sync your personal fork with the upstream repository.
> > To avoid this situation, create a new branch for your changes and reset the master branch. You can do this by running the following commands in a local repository for your personal fork. [...]
> 
> I am not sure whether I should still do this or not, since there is already some feedback, and I don't know what happens to that.
> In particular, I didn't understand if this is just a suggestion or actually required.
> 
> @aivanov-jdk if it's not required, do you have any preferences?

I see you opened the PR from master branch in your personal fork rather than a new branch.

Yes, I think you should go on as suggested: create a new PR from a new branch.

You can create the new PR with comments addressed and link to the old PR in the description.

In addition to that, if I get it right, you, @mperktold, have taken over this issue from @prsadhuk, thus #1628 is superseded by this PR and should be closed without integration. Eventually, this PR will be superseded by a new PR that you'll create.

Can you edit the subject of the JBS issue to amend the constant `W3C_LENGTH_UNITS`?

**Comment by mperktold** [comment]
> I see you opened the PR from master branch in your personal fork rather than a new branch.
> 
> Yes, I think you should go on as suggested: create a new PR from a new branch.

OK, will do. If you don't mind, I have a couple of noob questions, as I'm new to the whole process:

> You can create the new PR with comments addressed and link to the old PR in the description.

Is there anything special I need to do to have "comments addressed"? Should I first apply your patches, and then proceed with the new fork afterwards?

> Can you edit the subject of the JBS issue to amend the constant W3C_LENGTH_UNITS?

I don't think I have the rights to do this. I don't even have an JBS account, I submitted the bug through the https://bugs.java.com/bugdatabase/. I just recently signed the OCA, is that sufficient to create a JBS account?

**Comment by aivanov-jdk** [comment]
> > You can create the new PR with comments addressed and link to the old PR in the description.
> 
> Is there anything special I need to do to have "comments addressed"? Should I first apply your patches, and then proceed with the new fork afterwards?

Whatever you like. You can incorporate the fixes to comments in your initial commit to the new branch, or you can fix them after you create the new branch with your exiting changeset.


> > Can you edit the subject of the JBS issue to amend the constant W3C_LENGTH_UNITS?
> 
> I don't think I have the rights to do this. I don't even have an JBS account, I submitted the bug through the https://bugs.java.com/bugdatabase/. I just recently signed the OCA, is that sufficient to create a JBS account?

No, it is not sufficient.
I have amended the subject of the bug.

I wondered why I didn't see OCA request in the PR. You had gotten it before you created your first PR.

**Comment by mrserb** [comment]
> I believe we're talking about javadoc spec: its text has not changed.
> CSS spec is updated. Since you follow the rules of CSS 2.2, it's better to link to CSS 2.2.

No, I did not mean the JavaDoc, my question was - "what part of the CSS spec was updated"?

**Comment by prsadhuk** [comment]
@mperktold I guess this PR needs to be closed since this is from master branch. Please create a new PR from your fix branch with this latest changeset. The review will continue on top of that.

**Comment by aivanov-jdk** [comment]
> > I believe we're talking about javadoc spec: its text has not changed.
> > CSS spec is updated. Since you follow the rules of CSS 2.2, it's better to link to CSS 2.2.
> 
> No, I did not mean the JavaDoc, my question was - "what part of the CSS spec was updated"?

You're right. I looked closely at the text in CSS 2.1 and 2.2 and length definitions are the same. CSS 2.1 contains a warning that several sections have been updated. Yet this particular one seems to have remained the same.

Yet the fix is still required. Swing handles High DPI by scaling all UI therefore all length units factors should not depend on the screen resolution or print resolution, the changes in resolution are automatically handled by 2d transform.

However, changing the link to the CSS specification is optional. Shall we update it to the most recent one?

**Comment by kevinrushforth** [comment]
Looks like this PR still needs to be closed.

**Pull request closed without merging**